### PR TITLE
feat: ability to deprecate endpoints

### PR DESCRIPTION
**Pull request by krackerman**

### CODE REVIEW

File: src/logic/wrapper.js [hunk 1]
```diff
@@ -7,7 +8,16 @@ module.exports.Wrapper = ({ router, module }) => {
   const wrapFn = (identifier, params, ...args) => {
     assert([1, 2].includes(args.length));
     const [options, handler] = args.length === 2 ? args : [{}, args[0]];
+    // process.env.RATE_LIMIT is a string
```
Comment (krackerman) [inline src/logic/wrapper.js:11]: Should we strictly enforce rate limiting being a Number??

File: test/handler.js [hunk 2]
```diff
@@ -46,6 +46,10 @@ module.exports.exception = api.wrap('GET exception', [], { limit: null }, () =>
   throw Error('Some Exception');
 });
 
+module.exports.deprecation = api.wrap('GET deprecation', [], {
+  limit: null, deprecated: '1.0.1'
```
Comment (simlu) [inline test/handler.js:50]: limit is not needed here, or is it?

File: src/logic/wrapper.js (hunk 1 above)
Comment (simlu) [reply src/logic/wrapper.js:11]: I think we should enforce as an integer and handle it as needed

File: test/handler.js [hunk 3]
```diff
@@ -46,15 +46,26 @@ module.exports.exception = api.wrap('GET exception', [], { limit: null }, () =>
...
 module.exports.text = api
   .wrap('GET text', [], () => api.ApiResponse('some text', 200, { 'some-header': 123 }));
 
-module.exports.json = api
-  .wrap('GET json', [], { limit: process.env.RATE_LIMIT }, () => api.JsonResponse({ some: 'json' }));
+module.exports.json = api.wrap('GET json', [],
+  process.env.RATE_LIMIT === undefined
+    ? {}
+    : { limit: parseInt(process.env.RATE_LIMIT, 10) },
+  () => api.JsonResponse({ some: 'json' }));
```
Comment (simlu) [inline test/handler.js:60]: just declare limit as a variable at the top of the file... and then declare as undefined or as integer and pass that in here. Much cleaner

File: src/logic/wrapper.js [hunk 4]
```diff
@@ -7,7 +8,16 @@ module.exports.Wrapper = ({ router, module }) => {
...
-    assert(options instanceof Object && !Array.isArray(options));
+    Joi.assert(options, Joi.object().keys({
+      limit: Joi
+        .number()
+        .integer()
+        .min(0)
+        .max(Number.MAX_SAFE_INTEGER)
+        .allow(null)
+        .optional(),
+      deprecated: Joi.string().pattern(/[0-9]\.[0-9]\.[0-9]/).optional()
```
Comment (simlu) [inline src/logic/wrapper.js:19]: We have a version regex somewhere. Abstract and reuse?

File: test/handler.js [hunk 5]
```diff
@@ -29,6 +29,10 @@ const api = require('../src/index').Api({
   }
 });
 
+const limit = typeof process.env.RATE_LIMIT === 'string'
+  ? parseInt(process.env.RATE_LIMIT, 10)
+  : process.env.RATE_LIMIT;
```
Comment (simlu) [inline test/handler.js:34]: Process.env.ratelimit === undefined? Undefined : parse

**Pull request merged**